Revised the unifying techniques section (#26)
greenelab/knowledge-graph-review: content/04.network_applications.md

Knowledge graphs can help researchers tackle many biomedical tasks such as finding new treatments for existing drugs [@doi:10.7554/eLife.26726], aiding efforts to diagnose patients [@arxiv:1611.07012] and predicting associations between diseases and biomolecules [@doi:10.1155/2017/2498957].
In many cases, solutions rely on representing knowledges graphs in a low dimensional space, which is a process called representation learning.
This space preserves a knowledge graph's local and/or global structure and can support efforts to apply machine learning methods to make predictions.

> In the next few sections we discuss the unifying techniques that construct this low dimensional space and unifying applications that use this space to solve biomedical problems.

Reviewer: ```suggestion
In the next sections we review the unifying techniques that construct this low dimensional space and unifying applications that use this space to solve biomedical problems.
```

We group these methods into the following three categories: matrix factorization, translational methods, and deep learning (Figure {@fig:unifying_techniques_overview}).

![
Pipeline for representing knowledge graphs in a low dimensional space.
Starting with a knowledge graph, this space can be generated using one of the following options: Matrix Factorization (a), Translational Models (b) or Deep Learning (c).
The output of this pipeline is an embedding space that clusters similar node types together.
](images/figures/unifying_techniques_overview.png){#fig:unifying_techniques_overview}

#### Matrix Factorization

Matrix factorization is a technique that uses linear algebra to map high dimensional data onto a low dimensional space.
This projection is accomplished by decomposing a matrix into a set of small rectangular matrices (Figure {@fig:unifying_techniques_overview} (a)).
Notable methods for matrix decomposition include Isomap [@doi:10.1126/science.290.5500.2319], Laplacian eigenmaps [@doi:10.1162/089976603321780317] and Principal Component Analysis (PCA) [@doi:10/bm8dnf]/Singular Vector Decomposition (SVD) [@doi:10.1007/BF02288367].

> These methods were designed to be used on many different types of data; however, we focus on using them in the context of representing a knowledge graphs in a low dimensional space.	

Reviewer: ```suggestion
These methods were designed to be used on many different types of data; however, we discus their use in the context of representing a knowledge graphs in a low dimensional space.	
```

#### Translational Distance Models

Translational distance models treat edges in a knowledge graph as linear transformations.
For example, one such algorithm, TransE [@raw:bordestranse], treats every node-edge pair as a triplet with head nodes represented as $\textbf{h}$, edges represented as $\textbf{r}$, and tail nodes represented as $\textbf{t}$.
These representations are combined into an equation that mimics the iconic word vectors translations ($\textbf{king} - \textbf{man} + \textbf{woman} \approx \textbf{queen}$) from the word2vec model [@arxiv:1310.4546].
The described equation is shown as follows: $\textbf{h} + \textbf{r} \approx \textbf{t}$.
Starting at the head node ($\textbf{h}$), add the edge vector ($\textbf{r}$) and the result should be the tail node ($\textbf{t}$).
TransE optimizes vectors for $\textbf{h}$, $\textbf{r}$, $\textbf{t}$, while guaranteeing the global equation ($\textbf{h} + \textbf{r} \approx \textbf{t}$) is satisfied [@raw:bordestranse].
A caveat to the TransE approach is that it the training steps force relationships to have a one to one mapping, which may not be appropriate for all relationship types.

> Wang et al., attempted to resolve the one to one mapping issue by developing the TransH model [@raw:wangtransH].

Reviewer: ```suggestion
Wang et al. attempted to resolve the one to one mapping issue by developing the TransH model [@raw:wangtransH].
```

#### Deep Learning

Deep learning is a paradigm that uses multiple non-linear transformations to map high dimensional data into a low dimensional space.
Many techniques that use deep learning on knowledge graphs are based on word2vec [@arxiv:1310.4546; @arxiv:1301.3781], a set of approaches that are widely used for natural language processing.
The goal of word2vec is to project words onto a low dimensional space that preserves their semantic meaning.
Strategies for training word2vec models use one of two neural network architectures: skip-gram and continuous bag of words (CBOW).
Both models are feed-forward neural networks, but CBOW models are trained to predict a word given its context while skip-gram models are trained to predict the context given a word [@arxiv:1310.4546; @arxiv:1301.3781].
Once training has finished, words are now associated with dense vectors that downstream models, such as feed forward networks or recurrent networks, can use for input.

Deepwalk is an early method that represents knowledge graphs in a low dimensional space [@arxiv:1403.6652]. 
The first step of this method is to perform a random walk along a knowledge graph.

> During the random walk, every generated sequence of nodes is recorded and treated exactly similar to a sentence in word2vec [@arxiv:1310.4546; @arxiv:1301.3781].

Reviewer: ```suggestion
During the random walk, every generated sequence of nodes is recorded and treated as a sentence in word2vec [@arxiv:1310.4546; @arxiv:1301.3781].
```

These approaches use neural networks called autoencoders to generate this low dimensional space [@arxiv:1802.08352; @arxiv:1611.07308; @arxiv:1802.04407].
Autoencoders map input such as an adjacency matrices into a low dimensional space and then determines how to construct this space via reconstructing the same input [@arxiv:1404.7828; @raw:Baldi2011].
This generated space represents the nodes and their connectivity structure within a knowledge graph [@arxiv:1802.08352; @arxiv:1611.07308; @arxiv:1802.04407].
Despite the high potential of this approach, this method relies on an adjacency matrix for input which can run into scalability issues as a knowledge graph asymptotically increases in size [@arxiv:1903.07902].
Plus, Khosla et al., discovered that approaches akin to node2vec outperformed algorithms using autoencoders when undergoing link prediction and node classification [@arxiv:1903.07902].
Overall, the performance of deep learning techniques largely depends upon the structure of nodes and edges within a knowledge graph [@arxiv:1903.07902].

> Future efforts should consider creating hybrid models that use both node2vec and autoencoders to construct complementary low dimensional representations of knowledge graphs.

Reviewer: ```suggestion
Future work should include hybrid models that use both node2vec and autoencoders to construct complementary low dimensional representations of knowledge graphs.
```
